File: 0347-top-k-frequent-elements/0347-top-k-frequent-elements.java
class Solution {
    public int[] topKFrequent(int[] nums, int k) {
        int[] result = new int[k];
        HashMap<Integer, Integer> map = new HashMap<>();
        for(int i : nums){
            map.put(i, map.getOrDefault(i, 0) + 1);
        }
        PriorityQueue<Integer> pq = new PriorityQueue<>(
            (a, b) -> Integer.compare(map.get(b), map.get(a))
        );
        for(int key : map.keySet()){
            pq.add(key);
        }
        for(int i = 0; i < k; i++){
            result[i] = pq.poll();
        }
        return result;
    }
}

File: 0347-top-k-frequent-elements/0347-top-k-frequent-elements.py
class Solution:
    def topKFrequent(self, nums: list[int], k: int) -> list[int]:
        hashMap = {}
        for num in nums:
            hashMap[num] = hashMap.get(num, 0) + 1

        h = []
        for i in hashMap:
            h.append(hashMap.get(i, 0))
        maxHeap = [-n for n in h]
        heapq.heapify(maxHeap)

        result = []
        while len(result) < k:
            freq = -(heapq.heappop(maxHeap))
            for i in hashMap:
                if hashMap[i] == freq and i not in result:
                    result.append(i)
                    if len(result) == k:
                        break
        return result


                
